torchsde/_core/methods/srk.py

```diff
@@ -63,38 +59,31 @@ def diagonal_or_scalar_step(self, t0, t1, y0):
         dt = t1 - t0
         sqrt_dt = torch.sqrt(dt) if isinstance(dt, torch.Tensor) else math.sqrt(dt)
         I_k, I_k0 = self.bm(t0, t1, return_U=True)
-        I_kk = [(delta_bm_ ** 2. - dt) / 2. for delta_bm_ in I_k]
-        I_kkk = [(delta_bm_ ** 3. - 3. * dt * delta_bm_) / 6. for delta_bm_ in I_k]
+        I_kk = (I_k ** 2 - dt) / 2.
+        I_kkk = (I_k ** 3 - 3. * dt * I_k) / 6.
 
         y1 = y0
         H0, H1 = [], []
         for s in range(srid2.STAGES):
             H0s, H1s = y0, y0  # Values at the current stage to be accumulated.
             for j in range(s):
-                f_eval = self.sde.f(t0 + srid2.C0[j] * dt, H0[j])
-                g_eval = self.sde.g(t0 + srid2.C1[j] * dt, H1[j])
-                H0s = [
-                    H0s_ + srid2.A0[s][j] * f_eval_ * dt + srid2.B0[s][j] * g_eval_ * I_k0_ / dt
-                    for H0s_, f_eval_, g_eval_, I_k0_ in zip(H0s, f_eval, g_eval, I_k0)
-                ]
-                H1s = [
-                    H1s_ + srid2.A1[s][j] * f_eval_ * dt + srid2.B1[s][j] * g_eval_ * sqrt_dt
-                    for H1s_, f_eval_, g_eval_ in zip(H1s, f_eval, g_eval)
-                ]
+                f = self.sde.f(t0 + srid2.C0[j] * dt, H0[j])
+                g = self.sde.g(t0 + srid2.C1[j] * dt, H1[j])
+                g = g.squeeze(2) if g.dim() == 3 else g
```

Comment:
This is going to break if/when we support multiple batch dimensions. (Plus it's a bit opaque what it's doing.) Factor this out to a different method for additive/diagonal a la Milstein?

Reply:
I think we should just write wrappers at some point collapsing/uncollapsing the extra dimensions, since writing a single wrapper class is less work than changing all the solvers.

Reply:
Makes sense. We already try to factor as much of the noise type abstraction into the SDE (and out of the solvers) as it is, after all.